File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        front = 0
        triplets = []
        while front <= len(nums) - 3:
            mid = front+1
            back = len(nums) - 1
            target = -nums[front]
            while mid < back:
                if nums[mid] + nums[back] < target:
                    temp = nums[mid]
                    while mid < back and nums[mid] == temp: mid+=1
                elif nums[mid] + nums[back] > target:
                    temp = nums[back]
                    while mid < back and nums[back] == temp: back -= 1
                else:
                    triplets.append([nums[front], nums[mid], nums[back]])
                    temp = nums[mid]
                    while mid < back and nums[mid] == temp: mid+=1
                    temp = nums[back]
                    while mid < back and nums[back] == temp: back -= 1
            temp = nums[front]
            while front <= len(nums) - 3 and nums[front] == temp: front += 1
        return triplets

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut l, mut r) = (0usize, nums.len()-1);
        while nums[l] > nums[r] {
            let mid = (l+r) / 2;
            if nums[mid] < nums[l] {
                r = mid;
            } else {
                l = mid+1;
            }
        }

        if l == 0{
            r = nums.len() - 1;
        } else{
            r = l-1;
        }
        while r != l {
            let mid = if r > l {(r+l)/2} else {((r+nums.len() + l) / 2) % nums.len()};
            if nums[mid] < target {
                l = (mid+1)%nums.len();
            } else if nums[mid] > target{
                r = mid;
            } else {
                return mid.try_into().unwrap();
            }
        }

        if nums[l] == target{
            l.try_into().unwrap()
        } else{
            -1
        }
    }
}
